### PR TITLE
Tentative fix for #6520: camlcity.org unresponsive makes AppVeyor fail.

### LINKED ISSUE #6520: download.camlcity.org unresponsive, makes AppVeyor fail
camlcity.org is down (again) and this makes the packaging jobs on AppVeyor fail (because they download findlib from this site). Could we work around this untrustworthy website that hosts findlib?

### PR DESCRIPTION
**Pull request by Zimmi48**
**Kind:** infrastructure fix

Fixes / closes #6520 

Let's hope that this works! BTW, I don't like the many unsafe HTTP downloads that this script contains. Especially, given that we subsequently sign and distribute the resulting packages.

### CODE REVIEW

**Comment by Zimmi48** [comment]
Another option, apparently used by OPAM itself, is to use the mirror http://download2.camlcity.org/download/findlib-1.7.3.tar.gz when the main URL http://download.camlcity.org/download/findlib-1.7.3.tar.gz is down.

**Comment by Zimmi48** [comment]
Interesting, https://ci.appveyor.com/project/Zimmi48/coq/build/fix%2fcamlcity-down%7e285 now (apparently) blocks on the download of http://ftp.inf.tu-dresden.de/software/windows/cygwin32/x86_64/release/mingw64-i686-binutils/mingw64-i686-binutils-2.29.1.787c9873-1-src.tar.xz (which I just tested downloads instantly on my machine). It could be AppVeyor themselves having serious issues with their servers... (although they report no incidents). In any case, @maximedenes we are no lucky with CI and releases...

**Comment by Zimmi48** [comment]
It looks like the problems with AppVeyor have vanished once again so this PR becomes less urgent (something we can worry about after the 8.7.2 deadline).

**Comment by maximedenes** [comment]
@Zimmi48 Is there a way to try multiple locations? It would be more robust I guess.

**Comment by Zimmi48** [comment]
I agree. And no idea. We could ask @MSoegtropIMC if he knows.

**Comment by Zimmi48** [comment]
While this PR is likely to decrease the frequency of failures, I agree with @maximedenes though that it would be better if we could not rely on a single location but have a list of mirrors. That being said, we can merge this now and improve it later, given that this is currently making our CI fail. I'll close and reopen to confirm that this fixes the current problem.

**Comment by MSoegtropIMC** [comment]
I can add multiple mirror support with the hash checks. But in this specific case it wouldn't work because the files are not identical (see my post above).

Would anybody object if I supply a mirror space on one of my privately rented servers (preferably Amazon, but I can also offer Strato). The few € this costs me per year is probably well spent - I don't like systems breaking over the weekend.

Best regards,

Michael

**Comment by ejgallego** [comment]
camlcity already has its own mirror to avoid these problems, see: https://github.com/ocaml/opam-repository/blob/master/packages/ocamlfind/ocamlfind.1.7.3-1/url

**Comment by Zimmi48** [comment]
You are right @ejgallego. Let's try to use this mirror instead. That being said we can still merge this PR as an emergency fix, then backtrack when implementing the mirror functionality.

@MSoegtropIMC I personally wouldn't object to you setting up some mirrors (of course it may not be necessary in this case given @ejgallego's remark). Mirrors can only make things more robust, not less, and with the hash checks we don't even have to trust the mirrors.

**Comment by ejgallego** [comment]
OPAM already has support for mirrors, and the files are kept up to date. I still don't understand why the windows build system doesn't use OPAM, even if it is only to grab the sources.

**Comment by MSoegtropIMC** [comment]
@ejgallego when I started this, the Coq Windows build got stuff from quite a range of rather dubious sources (I don't mean opam with this - opam for Windows was just emerging back then). In order to clean this up, I tried to get everything from "original" sources, which I still think is the right approach - if only for documentation purposes / whom to contact if something doesn't work.

Another reason is that opam does some minor changes to the source archives (different root folder name, an extra file in the case of findlib) which makes it difficult to just hash the archive and tell it is identical to the original one.

But there is no strong reason. Currently the Windows build scripts get things from Cygwin and original sources (or what I think they are). One could opam to the list of "trusted sources", but I think the list of files we could get from opam would cover only a small fraction of what is needed. So I think (my personal opinion) we should keep the original sources and add one mirror for everything.

Best regards,

Michael

**Comment by Zimmi48** [comment]
Note that this PR already passed the 32-bit build step and it should pass the 64-bit build step in a few minutes (see https://ci.appveyor.com/project/coq/coq/build/master~3177), confirming that it solves the current problem. @MSoegtropIMC if you agree that this PR is OK as a temporary emergency fix, then as a code owner you can decide to merge this.

NB: normally all CI testing should pass before a PR is merged but this PR is an exception as it only touches the Windows build scripts and it is quite urgent. So in this case, it is OK to only check the result of AppVeyor.
(We can see coq-dpdgraph, equations, fiat-parsers and ltac2 failing but it is because these plugins received non-backward-compatible fixes between when this branch was prepared and now. Nothing to worry about.)

**Comment by MSoegtropIMC** [comment]
Since I didn't do a Coq merge before, I need to do a few things

- check if the merge script works on Cygwin (I haven't got a Linux machine here, at least not one with internet access, also possibly I need to add the required tools)

- check if my signature key known to GitHub is what I want to use long term for this purpose (likely not)

- possibly a few other things I find on the way

Since I have a few other things to do, this might take until 10 PM today. I am happy to look into this, but in case this is super urgent and 10 PM is too late, could you ask e.g. @maximedenes?

Btw.: I am not sure I already have merge rights ...

**Comment by MSoegtropIMC** [comment]
P.S.: I am generally fine with this a s a quick fix since it is only server outage related. I anyway plan to change this.

**Comment by Zimmi48** [comment]
> Btw.: I am not sure I already have merge rights ...

I have given them to you already ;-)

I suppose it's a good time to test the merge script on Cygwin.
To me it makes no difference if it is merged now or on 10 PM but @maximedenes feel free to merge this before if you want.

**Comment by MSoegtropIMC** [comment]
Ok, I give it a go then ...

Can you make me reviewer? Maxime seems to be busy ...

**Comment by maximedenes** [comment]
I can merge it now, if you like.

**Review by maximedenes — APPROVED** [review]

**Comment by MSoegtropIMC** [comment]
@maximedenes: I can also do it - just takes longer ... but it would be a good test if it can be done from Cygwin. In case I have issues, I will let you know.

**Comment by maximedenes** [comment]
Oops, sorry! I saw the last message too late. But maybe we can test the cygwin support with a less pressing patch, at your convenience @MSoegtropIMC.

**Comment by MSoegtropIMC** [comment]
@maximedenes: this is entirely your decision - I can't tell how urgent it is. If you need it now, please go ahead.
I will anyway set up a Cygwin with the required tools and can test it later.

**Pull request merged**
